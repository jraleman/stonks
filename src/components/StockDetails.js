import React, { useContext } from 'react';
import isEmpty from 'lodash.isempty';
import { StocksStateContext } from '../context/stocks-context';

const StockDetails = () => {
    const { symbol, charts } = useContext(StocksStateContext);
    if (!symbol || isEmpty(charts)) {
        return null;
    }
    return (
        <div>
            {JSON.stringify(charts)}
        </div>
    );
};

export default StockDetails;
