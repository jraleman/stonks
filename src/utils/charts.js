import * as d3 from 'd3';
import isEmpty from 'lodash.isempty';

// source: https://brendansudol.com/writing/responsive-d3
const responsivefy = (svg) => {
    if (!svg.node()) {
        return;
    }
    // get container + svg aspect ratio
    const container = d3.select(svg.node().parentNode);
    const width = parseInt(svg.style('width'));
    const height = parseInt(svg.style('height'));
    const aspect = width / height;

    // get width of container and resize svg to fit it
    const resize = () => {
        const targetWidth = parseInt(container.style('width'));
        svg.attr('width', targetWidth);
        svg.attr('height', Math.round(targetWidth / aspect));
    }

    // add viewBox and preserveAspectRatio properties,
    // and call resize so that svg resizes on inital page load
    svg.attr('viewBox', `0 0 ${width} ${height}`)
        .attr('perserveAspectRatio', 'xMinYMid')
        .call(resize);

    // to register multiple listeners for same event type, 
    // you need to add namespace, i.e., 'click.foo'
    // necessary if you call invoke this function for multiple svgs
    // api docs: https://github.com/mbostock/d3/wiki/Selections#on
    d3.select(window).on(`resize.${container.attr('id')}`, resize);
};

const movingAverage = (data, numberOfPricePoints) =>
    data.map((row, index, total) => {
        const start = Math.max(0, index - numberOfPricePoints);
        const subset = total.slice(start, index + 1);
        const sum = subset.reduce((a, b) => a + b.close, 0);
        return ({
            date: new Date(row.date),
            average: sum / subset.length
        });
    });

const clearChart = (id) => {
    d3.select(id).select('svg').remove();
};

export const drawLinearChart = ({
    id = '',
    data = [{}],
    xAxis = '',
    yAxis = '',
    xConvert = (x) => x,
    yConvert = (y) => y,
    colors = {
        price: '#4682b4',
        average: '#ffa500',
        gain: '#03a678',
        lose: '#c0392b',
    }
}) => {
    const margin = { top: 50, right: 50, bottom: 50, left: 50 };
    const width = window.innerWidth - margin.left - margin.right;
    const height = window.innerHeight - margin.top - margin.bottom;

    const xMin = d3.min(data, d => xConvert(d[xAxis]));
    const xMax = d3.max(data, d => xConvert(d[xAxis]));
    const yMin = d3.min(data, d => yConvert(d[yAxis]));
    const yMax = d3.max(data, d => yConvert(d[yAxis]));

    const xScale = d3
        .scaleTime()
        .domain([xMin, xMax])
        .range([0, width]);
    
    const yScale = d3
        .scaleLinear()
        .domain([yMin - 5, yMax])
        .range([height, 0]);
    
    const line = d3
        .line()
        .x(d => xScale(xConvert(d[xAxis])))
        .y(d => yScale(yConvert(d[yAxis])));
    
    const movingAverageData = movingAverage(data, 49);

    const movingAverageLine = d3
        .line()
        .x(d => xScale(xConvert(d[xAxis])))
        .y(d => yScale(yConvert(d.average)))
        .curve(d3.curveBasis);
    
    const volData = data.filter(d => d.volume);
    const yMinVolume = d3.min(volData, d => Math.min(d.volume));
    const yMaxVolume = d3.max(volData, d => Math.max(d.volume));
    const yVolumeScale = d3
        .scaleLinear()
        .domain([yMinVolume, yMaxVolume])
        .range([height, 0]);

    clearChart(id);

    // setup
    const svg = d3
        .select(id)
        .append('svg')
        .attr('width', width + margin.left + margin.right)
        .attr('height', height + margin.top + margin.bottom)
        .call(responsivefy)
        .append('g')
        .attr('transform', `translate(${margin.left},  ${margin.top})`);
    // x axis
    svg
        .append('g')
        .attr('id', 'xAxis')
        .attr('transform', `translate(0, ${height})`)
        .call(d3.axisBottom(xScale));
    // y axis
    svg
        .append('g')
        .attr('id', 'yAxis')
        .attr('transform', `translate(${width}, 0)`)
        .call(d3.axisRight(yScale));
    // draw price line
    svg
        .append('path')
        .data([data])
        .style('fill', 'none')
        .attr('id', 'priceChart')
        .attr('stroke', colors.price)
        .attr('stroke-width', '1.5')
        .attr('d', line);
    // draw avg curve
    svg
        .append('path')
        .data([movingAverageData])
        .style('fill', 'none')
        .attr('id', 'movingAverageLine')
        .attr('stroke', colors.average)
        .attr('d', movingAverageLine);
    // draw volume bar
    svg
        .selectAll()
        .data(volData)
        .enter()
        .append('rect')
        .attr('x', d => xScale(xConvert(d[xAxis])))
        .attr('y', d => yVolumeScale(d.volume))
        .attr('fill', (d, i) =>
            (i && volData[i - 1].close > d.close ? colors.lose : colors.gain))
        .attr('width', 1)
        .attr('height', d => (height - yVolumeScale(d.volume)));
    
    return svg;
};

export const getAverageInfo = (stockData) => {
    if (isEmpty(stockData)) {
        return [{}];
    }
    return (
        Array.from(stockData.reduce(
            (acc, obj) => Object.keys(obj).reduce( 
                (acc, key) => typeof obj[key] == "number"
                    ? acc.set(key, (acc.get(key) || []).concat(obj[key]))
                    : acc,
            acc),
            new Map()
        )).reduce(
            (acc, [name, values]) =>
                Object.assign(acc, { [name]: values.reduce((a,b) => a + b) / values.length }),
            {}
    ));
};
