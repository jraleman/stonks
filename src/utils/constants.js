export const navTabs = [
    { title: 'Stocks List', href: '/stocks' },
];
export const navDropdownItems = [
    { label: 'About', onClick: () => window.alert('about') },
    { label: 'Language', onClick: () => window.alert('language') },
];
export const baseURL = 'https://cloud.iexapis.com/stable';
