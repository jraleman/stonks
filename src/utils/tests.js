import { render } from '@testing-library/react';
import StocksProvider from '../context/stocks-context';

export const stocksCtxRender = (ui, { initialState = {}, ...renderOptions }) =>
    render(
        <StocksProvider {...initialState}>
            {ui}
        </StocksProvider>,
        renderOptions
    );
