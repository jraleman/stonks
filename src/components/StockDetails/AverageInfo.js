import React from 'react';
import { Alert } from 'reactstrap';
import uniqueId from 'lodash.uniqueid';
import styled from 'styled-components';
import { additionalInfo, averageInfoLabel } from '../../utils/constants';

const StyledInfo = styled.div`
    line-height: 2;
    margin-left: 2em;
    font-weight: 500;
`;

const InfoContainer = styled.div`
    background: #f9f9f9;
    color: #121212;
    width: 100%;
`;

const AverageInfo = ({
    info = [],
    allowed = additionalInfo
}) => (
    <InfoContainer>
        <Alert color="info">{averageInfoLabel}</Alert>
        { allowed.map((a) => info[a] && (
            <StyledInfo key={uniqueId('average-info_')}>
                {`â‡° ${a} = ${info[a].toFixed(2)}`}
            </StyledInfo>
        ))}
        <hr />
    </InfoContainer>
);

export default AverageInfo;
