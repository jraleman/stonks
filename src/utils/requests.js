import axios from "axios";
import { baseURL } from './constants';

const { REACT_APP_API_TOKEN: token } = process.env || {};

const api = axios.create({
    baseURL,
    params: { token }
});

export const getStatus = async () => {
    try {
        await api.get(`/status`);
        return true;
    } catch (error) {
        console.error(error);
        return false;
    }
};

export const getQuotesForStock = async ({ symbol = '' }) => {
    try {
        const res = await api.get(`/stock/${symbol}/quote`);
        return res;
    } catch (error) {
        const message = error || {};
        throw new Error(message);
    }
};

export const getLogoForStock = async ({ symbol = '' }) => {
    try {
        const res = await api.get(`/stock/${symbol}/logo`);
        return res;
    } catch (error) {
        const message = error || {};
        throw new Error(message);
    }
};

export const getChartForStock = async ({ symbol = '', range = '' }) => {
    try {
        const res = await api.get(`/stock/${symbol}/chart/${range}`);
        return res;
    } catch (error) {
        const message = error || {};
        throw new Error(message);
    }
};