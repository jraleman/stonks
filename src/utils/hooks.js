import { useState, useEffect } from 'react';
import { getQuotesForStock, getLogoForStock, getStatus } from './requests';
import { stocksList } from './constants';

// Returns stocks data, and status of server
export const useStocks = () => {
    const [status, setStatus] = useState(false);
    const [stocks, setStocks] = useState([]);

    useEffect(() => {
        const s = getStatus();
        setStatus(s);
    }, [setStatus]);

    useEffect(() => {
        const fetchQuote = async ({ symbol }) => {
            const { data: stock } = await getQuotesForStock({ symbol });
            const { data: { url: logo }} = await getLogoForStock({ symbol });
            const item = { ...stock, logo };
            setStocks((s) => [...s, item])
        };
        const populateStocks = async () => {
            for (let i = 0; i < stocksList.length; i += 1) {
                fetchQuote({ symbol: stocksList[i] });
            }
        }
        if (status) {
            populateStocks();
        }
    }, [status]);

    return [stocks, status];
};
