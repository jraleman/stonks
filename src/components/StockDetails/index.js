import React, { useContext } from 'react';
import isEmpty from 'lodash.isempty';
import styled from 'styled-components';
import { StocksStateContext } from '../../context/stocks-context';
import { getAverageInfo } from '../../utils/charts';
import { useLinearChart } from '../../utils/hooks';
import AverageInfo from './AverageInfo';

const Container = styled.div`
    padding: 1em;
`;

// get from theme context
const StyledChart = styled.div`
    background: #121212;
    color: #f9f9f9;
`;

const StockDetails = () => {
    const { symbol, stockData } = useContext(StocksStateContext);
    useLinearChart({ data: stockData });

    const avg = getAverageInfo(stockData);
    if (!symbol || isEmpty(stockData)) {
        return null;
    }
    return (
        <Container>
            <StyledChart id="chart" />
            <AverageInfo info={avg} />
        </Container>
    );
};

export default StockDetails;
