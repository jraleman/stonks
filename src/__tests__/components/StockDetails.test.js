import { render } from '@testing-library/react';
import StockDetails from '../../components/StockDetails';
import AverageInfo from '../../components/StockDetails/AverageInfo';
import { additionalInfo, averageInfoLabel } from '../../utils/constants';
import { getAverageInfo } from '../../utils/charts';
import stockData from '../../__mocks__/data/stockData';
import { stocksCtxRender } from '../../utils/tests';

describe('StockDetails', () => {
    it('should render', () => {
        const initialState = {
            symbol: '',
            stockData: [{}]
        };
        const component = stocksCtxRender(<StockDetails />, { initialState });
        expect(component).toMatchSnapshot();
    });

    it('uses stock data', () => {
        const initialState = {
            symbol: 'NKE',
            stockData
        };
        const component = stocksCtxRender(<StockDetails />, { initialState });
        expect(component).toMatchSnapshot();
    });
});

describe('AverageInfo', () => {
    it('should render', () => {
        const component = render(<AverageInfo />);
        expect(component).toMatchSnapshot();
    });

    it('uses stock data', () => {
        const avg = getAverageInfo(stockData);
        const component = render(<AverageInfo info={avg} allowed={additionalInfo} />);
        expect(component).toMatchSnapshot();
    });

    it('checks for label', () => {
        const { getByText } = render(<AverageInfo />);
        expect(getByText(averageInfoLabel)).toBeDefined();
    });
});
