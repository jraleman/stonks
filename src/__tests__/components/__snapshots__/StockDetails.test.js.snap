// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AverageInfo should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dkPtyc bYitdf"
      >
        <div
          class="alert alert-info fade"
          role="alert"
        >
          -- Additional Information --
        </div>
        <hr />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dkPtyc bYitdf"
    >
      <div
        class="alert alert-info fade"
        role="alert"
      >
        -- Additional Information --
      </div>
      <hr />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`AverageInfo uses stock data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dkPtyc bYitdf"
      >
        <div
          class="alert alert-info fade"
          role="alert"
        >
          -- Additional Information --
        </div>
        <div
          class="sc-gsDJrp feFTdM"
        >
          ⇰ high = 134.05
        </div>
        <div
          class="sc-gsDJrp feFTdM"
        >
          ⇰ low = 131.66
        </div>
        <div
          class="sc-gsDJrp feFTdM"
        >
          ⇰ close = 132.88
        </div>
        <div
          class="sc-gsDJrp feFTdM"
        >
          ⇰ volume = 6997675.64
        </div>
        <div
          class="sc-gsDJrp feFTdM"
        >
          ⇰ change = 0.01
        </div>
        <hr />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dkPtyc bYitdf"
    >
      <div
        class="alert alert-info fade"
        role="alert"
      >
        -- Additional Information --
      </div>
      <div
        class="sc-gsDJrp feFTdM"
      >
        ⇰ high = 134.05
      </div>
      <div
        class="sc-gsDJrp feFTdM"
      >
        ⇰ low = 131.66
      </div>
      <div
        class="sc-gsDJrp feFTdM"
      >
        ⇰ close = 132.88
      </div>
      <div
        class="sc-gsDJrp feFTdM"
      >
        ⇰ volume = 6997675.64
      </div>
      <div
        class="sc-gsDJrp feFTdM"
      >
        ⇰ change = 0.01
      </div>
      <hr />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`StockDetails should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`StockDetails uses stock data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
